service/history/shard/context_impl.go

```diff
@@ -633,11 +637,9 @@ func (s *ContextImpl) AddTasks(
 	}
 
 	s.wLock()
-	if err := s.errorByStateLocked(); err != nil {
-		s.wUnlock()
+	if err := s.errorByState(); err != nil {
```

Comment:
where is wUnlock? You removed it below?

Reply:
Is my understanding correct that here we have to check state within the shard lock otherwise the state may change after the check is done? Shard controller can operation on stale state information so only state lock is needed? 

If so, maybe worth adding some comments for it?

Reply:
> Is my understanding correct that here we have to check state within the shard lock otherwise the state may change after the check is done? Shard controller can operation on stale state information so only state lock is needed?

I don't think it's accurate to say that shard controller can operate on stale state information. The state is never stale. What may be stale is our view of ownership. But that's always the case: we can never be sure, no matter how many locks we hold, that the db and/or membership has not given the rangeid lock to someone else and we just didn't learn it yet.

The `errorByState` check is done so that 1. if we are unloading, we return an error immediately, 2. if we are trying to acquire, we don't do other ops until that completes successfully.

Should we check `errorByState` before or after the shard rwlock? I think after. If we check before, and then we block on acquiring the rwlock because there is another persistence op in progress, then we finally get the lock, it's possible that other op is the one where we learned that we lost the rangeid lock. So the state changed and we would use stale information. If we check after, at least we know that the previous op succeeded.

Can shard controller just check the state + membership and not worry about the actual ownership in the db? That's the current behavior, and this PR doesn't change it (previously it used the rwlock but it didn't do any additional persistence op, so effectively it was just checking the in-memory state). Wenquan suggested in #3135 that we do actually check with persistence, which seems like a good idea. But for another PR.

> If so, maybe worth adding some comments for it?

I can put some of the above in a comment. Do you have a suggestion for where the comment should go?